lib/elastic_apm/transport/base.rb

```diff
@@ -1,60 +1,52 @@
 # frozen_string_literal: true
 
 require 'elastic_apm/transport/connection'
-require 'elastic_apm/transport/serializers'
-require 'elastic_apm/transport/filters'
-
-require 'elastic_apm/metadata/service_info'
-require 'elastic_apm/metadata/system_info'
-require 'elastic_apm/metadata/process_info'
+require 'elastic_apm/transport/worker'
 
 module ElasticAPM
   module Transport
     class UnrecognizedResource < InternalError; end
 
     # @api private
     class Base
+      include Logging
+
       def initialize(config)
         @config = config
-
         @connection = Connection.new(config)
+        @worker = Worker.new(config, connection)
+      end
 
-        @serializers = Struct.new(:transaction, :span, :error).new(
-          Serializers::TransactionSerializer.new(config),
-          Serializers::SpanSerializer.new(config),
-          Serializers::ErrorSerializer.new(config)
-        )
+      attr_reader :config, :filters, :connection, :worker
 
-        @filters = Filters.new(config)
+      def submit(resource)
+        ensure_worker
+        worker.push(resource)
       end
 
-      attr_reader :config, :filters, :connection
+      def flush
+        connection.flush
+      end
 
-      # rubocop:disable Metrics/MethodLength
-      def submit(resource)
-        serialized =
-          case resource
-          when Transaction
-            @serializers.transaction.build(resource)
-          when Span
-            @serializers.span.build(resource)
-          when Error
-            @serializers.error.build(resource)
-          else
-            raise UnrecognizedResource
-          end
-
-        post serialized
+      def stop
+        stop_worker
+        flush
       end
-      # rubocop:enable Metrics/MethodLength
 
-      def post(payload)
-        @filters.apply payload
-        @connection.write(payload.to_json) unless config.disable_send?
+      private
+
+      def stop_worker
+        return unless @worker.alive?
+
+        debug 'Stopping worker'
+        @worker.stop 5
       end
 
-      def flush
-        @connection.close!
+      def ensure_worker
+        return if @worker.alive?
+
+        debug 'Starting worker'
+        @worker.work_forever
```

Comment:
This doesn't look thread safe and could start multiple workers if called by more than one thread.